src/server/builder.ts

```diff
@@ -0,0 +1,405 @@
+/// <reference path="..\compiler\commandLineParser.ts" />
+/// <reference path="..\services\services.ts" />
+/// <reference path="protocol.d.ts" />
+/// <reference path="session.ts" />
+/// <reference types="node" />
+
+namespace ts.server {
+
+    const crypto: any = require("crypto");
+
+    /**
+     * An abstract file info that maintains a shape signature.
+     */
+    export class BuilderFileInfo {
+
+        private lastCheckedShapeSignature: string;
+        private lastEmittedContentSignature: string;
+        scriptInfo: ScriptInfo;
+
+        constructor(public fileName: string) {
+        }
+
+        public isExternalModule(builder: Builder) {
+            const sourceFile = this.getSourceFile(builder);
+            if (!!sourceFile.externalModuleIndicator) {
+                return true;
+            }
+            if (sourceFile.isDeclarationFile) {
+                for (const statement of sourceFile.statements) {
+                    if (statement.kind !== SyntaxKind.ModuleDeclaration || (<ModuleDeclaration>statement).name.kind !== SyntaxKind.StringLiteral) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
+
+        public computeHash(text: string): string {
+            return crypto.createHash("md5")
+                .update(text)
+                .digest("base64");
+        }
+
+        public getSourceFile(builder: Builder): SourceFile {
+            return builder.project.getSourceFile(this.fileName);
+        }
+
+        public getScriptInfo(builder: Builder): ScriptInfo {
+            if (!this.scriptInfo) {
+                this.scriptInfo = builder.project.getScriptInfo(this.fileName);
+            }
+            return this.scriptInfo;
+        }
+
+        /**
+         * Thie method should only be called upon emitting
+         */
+        public updateLastEmittedContentSignature(builder: Builder) {
+            this.lastEmittedContentSignature = this.getScriptInfo(builder).getLatestVersion();
+        }
+
+        public checkIfShapeSignatureChanged(builder: Builder) {
+            const sourceFile = builder.project.getSourceFile(this.fileName);
+            if (!sourceFile) {
+                return true;
+            }
+
+            const lastSignature = this.lastCheckedShapeSignature;
+            if (sourceFile.isDeclarationFile) {
+                this.lastCheckedShapeSignature = this.computeHash(sourceFile.text);
+            }
+            const emitOutput = builder.project.getLanguageService(/*ensureSynchronized*/true).getEmitOutput(this.fileName, /*emitDeclarationsOnly*/ true);
+            for (const file of emitOutput.outputFiles) {
+                if (/\.d\.ts$/.test(file.name)) {

```

Comment:
is this check necessary given that you've requested to emit only declarations? Also having a loop is confusing since you should get output only for one file